rework the target page

Nexucis opened the pull request:
This PR aims to fix  #6881

It introduced a search bar and an infinite scroll per scrapePool.

For the moment, I implemented the fix in different files, but before merging this PR, it will entirely replace the current implementation of the Target page.

Edit:
It also fix #5219

Signed-off-by: Augustin Husson <husson.augustin@gmail.com>

<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolList2.tsx:
```diff
@@ -0,0 +1,131 @@
+      <Filter filter={filter} setFilter={setFilter} expanded={expanded} setExpanded={setExpanded} />
+      <Container>
+        <InputGroup>
+          <InputGroupText>{<FontAwesomeIcon icon={faSearch} />}</InputGroupText>
```
Could you wrap this in a prepend element like in https://github.com/prometheus/prometheus/blob/a961062c37f1675d5925d8b74ddda0a6322ed82d/web/ui/react-app/src/pages/graph/ExpressionInput.tsx#L219-L223, so it visually gets merged with the input?

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolContent.tsx:
```diff
@@ -0,0 +1,89 @@
+          {items.map((target, index) => {
+            return (
+              <tr key={index}>
+                <td>{target.scrapeUrl}</td>
```
These used to be links (via `EndpointLink`), now not anymore?

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolList2.tsx:
```diff
@@ -0,0 +1,131 @@
+  toggleExpanded: () => void;
+}
+
+function ScrapePoolPanel(props: PanelProps): JSX.Element {
```
Could you keep this in a separate file as it was before? We always put components in their own files unless one is really just a totally dumb wrapper around another.

I'd also remove the old implementations and just replace them with the new ones vs. having new ones with a `2` appended.

Nexucis replied:
in this case, I think it's a quite dumb wrapper. But if you really want I can move it in a different file of course.

juliusv reviewed (commented):
Looks like a good start, thanks! This is just a first round of comments, I'll take another more detailed view again (also about edge cases, UX, etc.) after another iteration.

I guess the individual panel list tables all need to have a maximum height (with scroller) now and can't expand to their full internal height anymore?

Also, should it already be possible to type in any label matchers like `instance="..."` or so (which doesn't seem to work yet), or is that planned for later only?

Nexucis commented:
Cool ! thanks for the first review

> I guess the individual panel list tables all need to have a maximum height (with scroller) now and can't expand to their full internal height anymore?

yes indeed, I tried to avoid to have individual scroller, but so far I didn't find a way to do it. The lib I used doesn't seem to handle it properly. But I think it's normal it doesn't handle it.

The maximum size of the table is defined like that: `height={items.length > 25 ? '75vh' : ''}`

> Also, should it already be possible to type in any label matchers like instance="..." or so (which doesn't seem to work yet), or is that planned for later only?

I did it through codemirror for the moment, but it's not yet available. After we talked around this topic, I was wondering if it will be needed. And instead I simplified the search by allowing user to just typing a word.

So currently when you are typing something, it will match the `scrapePoolName`, any label key or any label value.
Which should correspond more or less of what you are doing with `CTRL + F`

Nexucis commented on web/ui/react-app/src/pages/targets/ScrapePoolContent.tsx:
```diff
@@ -0,0 +1,89 @@
+        </thead>
+        <tbody>
+          {items.map((target, index) => {
+            return (
```
for this one, I think I didn't get what you mean.

I tried like that: 

```typescript
items.map((target, index) => {
   <tr key={index}>
   </tr>
}
```
but it doesn't look like it works.

juliusv replied:
You have to omit the `{ ... }` for that.

Nexucis commented on web/ui/react-app/src/pages/targets/ScrapePoolContent.tsx:
```diff
@@ -0,0 +1,89 @@
+  targets: Target[];
+}
+
+export function ScrapePoolContent(props: ScrapePoolContentProps): JSX.Element {
```
done for all function

Nexucis commented:
Alslo,  the tests currently failed with this message:

```
Cannot find module '@nexucis/kvsearch'
``` 

which is weird. I don't get why jest doesn't find the package.

Nexucis commented:
> Also, should it already be possible to type in any label matchers like instance="..." or so (which doesn't seem to work yet), or is that planned for later only?

I found this issue talking about this topic: https://github.com/prometheus/prometheus/issues/5532

For the moment with codemirror, you could write something like this `labels.instance=...`. Perhaps it can be enough ?
I didn't publish the `codemirror` part yet as I wasn't sure it would be useful for prometheus, and I didn't try if the autocompletion is efficiant with a huge pool of target.

But if you want to give a try, I can publish it and add it to this PR. In that case, do we want two different input ? An "expert" one with codemirror and a simple one like it is currently proposed in this PR. (Probably with a switch button to move from a version to another).

Nexucis commented:
alright I fixed the issue with jest and the fact it wasn't able to find the module `@nexucis/kvsearch`.

I also aligned the search bar and the filter buttons. At least it's cleaner like that

Nexucis commented:
> For the moment with codemirror, you could write something like this labels.instance=.... Perhaps it can be enough ?
I didn't publish the codemirror part yet as I wasn't sure it would be useful for prometheus, and I didn't try if the autocompletion is efficiant with a huge pool of target.

Here an example of what I have today with codemirror:

https://user-images.githubusercontent.com/4548045/148199454-4ff8e420-e8c5-42bc-83c6-2f25476154c7.mp4

beorn7 commented:
Reminder: In a week (on 2022-01-12), I'll cut a new release (v2.33). You could either try to expedite this enough to go into the next release. Or, on the other hand, if you feel this is too invasive to go into a release without sitting a while in main, you might want to delay the merge.

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolList.tsx:
```diff
@@ -1,61 +1,127 @@
+            <InputGroupAddon addonType="prepend">
+              <InputGroupText>{<FontAwesomeIcon icon={faSearch} />}</InputGroupText>
+            </InputGroupAddon>
+            <Input onChange={handleSearchChange} />
```
Could you add a placeholder like:

```suggestion
            <Input onChange={handleSearchChange} placeholder="Filter by endpoint or labels"/>
```

juliusv replied:
Maybe also add `autoFocus`?

Nexucis replied:
good idea ! Done :)

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolList.tsx:
```diff
@@ -1,61 +1,127 @@
+      setPoolList(
+        groupTargets(
+          result.map((value) => {
+            return value.original as unknown as Target;
```
Do you need two `as...` here?

Nexucis replied:
yeah, Typescript is yelling if I'm casting it directly to a Target.

juliusv commented on web/ui/react-app/src/pages/targets/target.ts:
```diff
@@ -2,7 +2,7 @@ export interface Labels {
 }
 
-export interface Target {
+export type Target = {
```
Is there a specific reason for changing from `type` to `interface` here?

juliusv replied:
Ok I read some more articles on `type` vs. `interface` and there's indeed very few differences nowadays. I'm only mildly concerned about the inconsistency of having `type Target`, but `interface DroppedTarget`, etc. Now `Target` is the odd one out. Not sure if it's worth keeping it an interface though if it introduces more hassle somewhere else.

Nexucis replied:
I don't really have an opinion on this topic. It's quite odd for me you have two differents ways to define an object.
I basically changed it because it was conveniant, but is it a really good idea, that's good question :')

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolContent.tsx:
```diff
@@ -0,0 +1,91 @@
+      dataLength={items.length}
+      height={items.length > 25 ? '75vh' : ''}
+    >
+      <Table className={styles.table} size="sm" hover>
```
In general with this new table design, it's good by itself (though striped could still be good for readability of long lines), but we have other pages with similar tables that still use the old design. So we either have to change all of them at once or keep this one at the old design for now.

Nexucis replied:
mmm let's keep the old design, there are already enough changes in this PR.

juliusv commented:
> Reminder: In a week (on 2022-01-12), I'll cut a new release (v2.33). You could either try to expedite this enough to go into the next release. Or, on the other hand, if you feel this is too invasive to go into a release without sitting a while in main, you might want to delay the merge.

@beorn7 I think it's fine to get it in (and it's almost done), that's what we have RCs for :)

juliusv commented:
From my side I'm pretty happy with it now and it performs super well even with 10k targets :partying_face: 

I'd love to get one more opinion on this from the UX perspective before merging, maybe @roidelapluie (since you are always very thoughtful :)). The scrolling-per-scrapepool behavior is maybe something I'd like to get a second opinion on the most.

Here's a quick Python (2) script to generate a Prometheus config with many dummy targets (in this example, 10k targets spread over 20 scrape configs):

```python
num_scrape_configs = 20
targets_per_scrape_config = 500

print "scrape_configs:"
for i in range(0, num_scrape_configs):
  print "- job_name: demo-%d" % i
  print "  static_configs:"
  print "  - targets:"
  for j in range(0, targets_per_scrape_config):
    print "    - foobar-host-%d:%d" % (i, j)
```

```bash
python2 gen-targets.py > prometheus.yml
```

Nexucis commented:
I added a short fix for the position of the tooltip for the labels. Without this fix, often the tooltip was displayed under the cursor and it had then a weird behavior.

In the meantime I fixed https://github.com/prometheus/prometheus/issues/5219

juliusv commented on web/ui/react-app/src/pages/targets/ScrapePoolList.tsx:
```diff
@@ -1,61 +1,140 @@
+
+  useEffect(() => {
+    const list = targetList.filter((t) => {
+      if (!showHealthy) {
```
Would be more natural to use `&&` vs. two nested `if` conditions, no?

Or just `filter((t) => showHealthy || t.health.toLowerCase() !== 'up'`

Nexucis replied:
ah yeah I don't know what I was thinking when coding this condition. Good eyes thanks !

juliusv commented:
Looks good to me besides my last nit. If nobody else complains about it, I'd merge it before the next release candidate to see if anyone complains about it then :)
